### ISSUE TITLE
Add new language

### ISSUE THREAD

**Issue by Mahabmn** [hulop/ConversationServiceWatson#2]
hello 

is this application still working ? 

how to add new language ? 

should we deploy it in a server ?

**Comment by daisukes** [comment]
Hello @Mahabmn 

Yes, this application (proxy server of Watson Assistant) should be working.

You can add a new language by adding configuration here.
https://github.com/hulop/ConversationServiceWatson/tree/master/ConversationServiceWatson/src/data


1. add some fixed messages for your language like others
https://github.com/hulop/ConversationServiceWatson/tree/master/ConversationServiceWatson/src/data/messages

2. create your Watson assistant workspace
https://www.ibm.com/products/watson-assistant

3. configure Watson Assistant workspace IDs here
https://github.com/hulop/ConversationServiceWatson/blob/5638aac9b71f7f557846b4cfec46d5b31924a2e5/ConversationServiceWatson/src/data/config.json#L9-L12

**Comment by Mahabmn** [comment]
thank you for the feedback , after updated the application and adding my new language should I deploy it in my server ?

**Comment by daisukes** [comment]
Yes, this is a Web application running on Java application server.

**Comment by Mahabmn** [comment]
thank  you , I will try to update it .

**Comment by Mahabmn** [comment]
can I test the application alone , or I should integrate it to NavCog then test it ?

**Comment by daisukes** [comment]
You can test the server without the NavCog app. Still, you need to prepare the IBM Watson Assistant workspace to which this proxy server connects.
Then, you can test the proxy server and the workspace by sending conversation queries to the server directly. ([API](https://cloud.ibm.com/apidocs/assistant-v1#message))

@itohtaka @kentarou-fukuda 
Do you have any good ideas to test this server app?

**Comment by Mahabmn** [comment]
what is the assistant us V1 or , V2 ? 

in the code is requesting assistant ID which is not available in IBM page , is this old parameter ?

**Comment by daisukes** [comment]
I think you may need to use V1 but it looks like the Web interface provides only V2.
I am not sure, but you may be able to create a V1 instance from CLI.

@kentarou-fukuda 
Could you help us?

**Comment by kentarou-fukuda** [comment]
@Mahabmn 
You can check your "Assistant ID" from "View API Details" menu of Watson Assistant. 

![image](https://user-images.githubusercontent.com/6054912/178288990-67a24c33-0b6b-4bdb-9c7d-12a11ceb76ac.png)

It covers both of Wattson Assistant v1 and v2 API.

Please check [Watson Assistant introduction](https://cloud.ibm.com/docs/watson-assistant) for more details.

**Comment by Mahabmn** [comment]
thank you 
should I add the entities , intents , dialogs to the Watson assistant page or to the code ?

**Comment by kentarou-fukuda** [comment]
Yes. To update the conversation, you need to edit entities, intents, dialogs in Watson Assisntant page.
Please see [simple example](https://github.com/hulop/ConversationServiceWatson/tree/master/SampleConversation). It might be a good starting point.

**Comment by Mahabmn** [comment]
hello , thank you , all parameters are clear 

but what should I add in dest_info ?

**Comment by Mahabmn** [comment]
hi 

did you use a json array proprieties in getting the feedback from the other servers ?

**Comment by Mahabmn** [comment]
hi 

what should we add in dest_info? 

how to connect the app with the map server ?

**Comment by daisukes** [comment]
https://github.com/hulop/HLPDialog/blob/a2dd24cbaf56a3d8c32b89f2e3c41f1fc6f84691/HLPDialog/DialogViewController.swift#L642

`dest_info` should have destination information. If your conversation ends up with, for example, Room A, then the `dest_info.nodes` is the node ID of the Room A on the map server.

`nodes` can have multiple node ids   separated by `|` to find the nearest node. It is useful to find the nearest bathroom for example.

**Comment by Mahabmn** [comment]
hello 

I didn't get the point 

did you mean I didn't have to connect the conversation service to map server ? 

you mean I have to update the dialog view controller with location information ? 

thank you

**Comment by Mahabmn** [comment]
hi , in Watson assistant page the 

dest_info is updated as "{"name":"@Destination","pron":"","nodes":"<? $node_map[@Destination] ?>","pr_short":""}"

this update is from the example of the conversation .

**Comment by daisukes** [comment]
Each server works independently. There is no function to utilize data on another server.
You need to bring node IDs from the MapService server to the conversation service.
As shown in the example, there is a node that sets up the `node_map` dictionary to have name-node mapping. Then you can refer to them as `"nodes": "<? $node_map[@Destination] ?>"`

```json
    {
      "go_to": null,
      "output": {
        "text": {
          "values": [
            "Let me start navigation to the <? @Destination ?>."
          ],
          "selection_policy": "sequential"
        }
      },
      "parent": null,
      "context": {
        "dest": "@Destination",
        "navi": true,
        "dest_info": {
          "name": "@Destination",
          "pron": "",
          "nodes": "<? $node_map[@Destination] ?>",
          "pr_short": ""
        }
      },
      "created": "2017-02-03T09:50:24.138Z",
      "updated": "2017-02-03T11:02:29.276Z",
      "metadata": null,
      "conditions": "@Destination",
      "description": null,
      "dialog_node": "node_5_1486115817438",
      "previous_sibling": "node_1_1486114964410"
    },
    {
      "go_to": {
        "return": null,
        "selector": "condition",
        "dialog_node": "node_1_1486114964410"
      },
      "output": {
      },
      "parent": null,
      "context": {
        "dest": "",
        "navi": false,
        "node_map": {
          "Toilet": "EDITOR_node_1486112611272",
          "Entrance": "EDITOR_node_1486112656238",
          "Auditorium": "EDITOR_node_1486112601650"
        },
        "dest_info": {
          "name": "",
          "pron": "",
          "nodes": "",
          "pr_short": ""
        },
        "agent_name": "HULOP"
      },
      "created": "2017-02-03T09:56:28.139Z",
      "updated": "2017-03-06T02:24:02.313Z",
      "metadata": null,
      "conditions": "true",
      "description": null,
      "dialog_node": "init",
      "previous_sibling": null
    },
```

**Comment by Mahabmn** [comment]
hi 

I use the same example but I didn't receive any last feedback . 

<img width="365" alt="Screen Shot 2022-07-22 at 3 23 54 PM" src="https://user-images.githubusercontent.com/99026115/180440374-ea24323b-cc07-42e1-84fa-4616a0c3e566.png">

**Comment by kentarou-fukuda** [comment]
Hi,

"Try it" in Watson Assistant UI is a kind of simulation. The sample conversation is aimed to use with ConversationServiceWatson application. To complete the conversation, you need to use it with ConversationServiceWatson application (and NavCog / MapService).
For example, if you try this conversation from NavCov app, the app will start navigation after "Let me start navigation to the Entrance" message.

**Comment by Mahabmn** [comment]
ok , thank you 

I will test it with the app .

**Comment by Mahabmn** [comment]
hello 

what should I add on the endpoint section 


		"endpoint": "https://gateway.watsonplatform.net/conversation/api/v1/workspaces/%s/message?version=2016-09-20",
		
	"logging_config": 
	{
		"endpoint": "$(CONV_MAP_SERVICE)/api/log",
	
thank you

**Comment by Mahabmn** [comment]
hi 

what I should add on this fields 

	},
	"logging_config": 
	{
		"endpoint": "$(CONV_MAP_SERVICE)/api/log",
		"api_key": "$(CONV_MAP_APIKEY)"
	}

**Comment by daisukes** [comment]
You don't need to change the file itself but configure some environment variables.

- `CONV_MAP_SERVICE`: your MapService hostname
- `CONV_MAP_APIKEY`: the key which you specified for the MapService as `AUDITOR_API_KEY`

The conversation system will store some logs on the MapService server.
This is an optional setting.

**Comment by daisukes** [comment]
Hi @Mahabmn,

I found that IBM Watson Assistant does not have V1 anymore, and it is version 2.
So, you need to change the endpoint.

https://github.com/hulop/ConversationServiceWatson/blob/5638aac9b71f7f557846b4cfec46d5b31924a2e5/ConversationServiceWatson/src/data/config.json#L5

```
https://api.us-south.assistant.watson.cloud.ibm.com/instances/<instance id>/v2/assistants/%s/message?version=2021-06-14
```

Please check your `<instance id>` at Watson assistant credential (URL).
![image](https://user-images.githubusercontent.com/43101027/193442133-a8af57f2-45d8-4d88-a48b-ead59b28e69e.png)


And Workspace ID (Assistatnt ID) can be found on Watson Assistant page.
![image](https://user-images.githubusercontent.com/43101027/193442219-8bf76513-d9f2-4ef8-aa6e-709217f05a0d.png)

**Comment by Mahabmn** [comment]
thank you @daisukes , Will try it and update you .

**Comment by Mahabmn** [comment]
hello @daisukes 

try to update the config file with the new parameter , but still not working . 

is there any other parameter need to change in the source code ? 

Thank you
